Fix #375: Capture plugin properties at validation time to eliminate timing mismatch

cowwoc opened the pull request:
## Problem

Maven 4 automatically injects `--module-version ${project.version}` into compiler arguments **during execution**, but the build cache validates properties **before execution**. This creates a timing mismatch that causes cache invalidation:

- **First build**: No cache exists → properties captured during/after execution (WITH Maven 4 injection)
- **Second build**: Cache exists → properties captured during validation (WITHOUT injection yet) → mismatch → cache invalidated

## Root Cause Analysis

The cache extension currently reads plugin properties at **different lifecycle points** for different builds:

```
First Build Timeline:
  1. findCachedBuild() → no cache found
  2. Mojos execute → Maven 4 injects --module-version
  3. beforeMojoExecution() fires → captures properties WITH injection
  4. save() stores properties from execution events

Second Build Timeline:
  1. findCachedBuild() → cache found
  2. verifyCacheConsistency() → creates mojo via getConfiguredMojo()
  3. Reads properties → WITHOUT injection (too early in lifecycle)
  4. Compares to first build's stored values (WITH injection)
  5. MISMATCH → cache invalidated!
```

The problem: **validation reads BEFORE injection, storage reads AFTER injection**.

## Solution

Capture properties at **validation time** for ALL builds (even when no cache exists). Store ONLY validation-time values in the cache. This ensures both validation and storage read at the same lifecycle point.

### Implementation

1. **Modified `CacheResult`** to store validation-time mojo events
2. **Modified `BuildCacheMojosExecutionStrategy`** to capture properties immediately after `findCachedBuild()`
3. **Modified `save()`** to store ONLY validation-time events (fails with `AssertionError` if missing)

### Key Code Changes

```java
// In BuildCacheMojosExecutionStrategy.execute():
result = cacheController.findCachedBuild(session, project, mojoExecutions, skipCache);

// NEW: Always capture validation-time properties when cache is enabled
if (cacheState == INITIALIZED) {
    Map<String, MojoExecutionEvent> validationTimeEvents =
        captureValidationTimeProperties(session, project, mojoExecutions);
    result = CacheResult.rebuilded(result, validationTimeEvents);
}
```

```java
// In save():
// Validation-time events MUST exist - no fallback to execution-time
if (result.getValidationTimeEvents() == null || result.getValidationTimeEvents().isEmpty()) {
    throw new AssertionError("Validation-time properties not captured - this is a bug");
}
cacheController.save(result, mojoExecutions, result.getValidationTimeEvents());
```

**Note:** Execution-time values are still captured by `MojoParametersListener` for logging/debugging during the build, but are NOT stored in the cache.

### New Timeline (Both Builds)

```
First Build:
  1. findCachedBuild() → no cache
  2. captureValidationTimeProperties() → reads WITHOUT injection
  3. Mojos execute → Maven 4 injects (logged but not stored)
  4. save() stores ONLY validation-time properties → WITHOUT injection

Second Build:
  1. findCachedBuild() → cache found
  2. verifyCacheConsistency() → reads WITHOUT injection
  3. captureValidationTimeProperties() → reads WITHOUT injection
  4. Compares to first build → BOTH without injection → MATCH!
```

### What Gets Stored vs. Logged

| Data | Stored in Cache | Available in Logs |
|------|-----------------|-------------------|
| Validation-time properties (no injection) | ✅ Yes | ✅ Yes |
| Execution-time properties (with injection) | ❌ No | ✅ Yes |

This approach ensures cache consistency while preserving debugging information in build logs.

## Benefits Over PR #391

| Aspect | PR #391 (`ignorePattern`) | This PR (Validation-Time Capture) |
|--------|--------------------------|-----------------------------------|
| **Approach** | Workaround: Filter mismatched values | Fix: Eliminate timing mismatch |
| **Configuration** | Required | None needed |
| **Flexibility** | Pattern must match version format | Format-agnostic |
| **Maintenance** | Patterns need updates | No maintenance |
| **Scope** | Only fixes known patterns | Fixes ALL auto-injected properties |

## Testing

Created comprehensive integration tests:

1. **Maven4JpmsModuleTest** - JPMS module with Maven 4 auto-injection of `--module-version`
2. **ExplicitModuleVersionTest** - JPMS module with explicit `moduleVersion` configuration
3. **NonJpmsProjectTest** - Regular Java project without JPMS (ensures no regression)
4. **MultiModuleJpmsTest** - Multi-module project with mixed JPMS and non-JPMS modules

All tests verify:
- ✅ First build creates cache
- ✅ Second build restores from cache (NO cache invalidation)
- ✅ NO `ignorePattern` configuration required

## Backward Compatibility

- Fully backward compatible
- Existing caches remain valid
- No configuration changes required
- `ignorePattern` still works but is no longer necessary for Maven 4 injection

## Alternative Approaches Considered

1. **Delay validation until execution** ❌ - Would lose early cache-miss detection performance benefit
2. **Pattern-based filtering (PR #391)** ❌ - Treats symptom (value mismatch) instead of root cause (timing mismatch)
3. **Maven core changes** ❌ - Outside our control, would take years to deploy
4. **Store both validation-time and execution-time** ❌ - Unnecessary complexity, doesn't solve the consistency problem
5. **Validation-time capture + storage (this PR)** ✅ - Fixes root cause, no configuration needed, preserves logging for debugging

## Related Issues

- Fixes #375
- Alternative to #391
- Related to Maven 4 JPMS module compilation

## Migration

No migration needed. This change is transparent to users. Existing projects will benefit automatically.

cowwoc committed:
Fix #375: Capture plugin properties at validation time to eliminate timing mismatch

Maven 4 automatically injects --module-version into compiler arguments during
execution, but cache validation happens before execution. This creates a timing
mismatch that causes cache invalidation.

This fix captures properties at validation time for ALL builds and stores ONLY
validation-time values in the cache, ensuring consistent reading at the same
lifecycle point.

Changes:
- Modified CacheResult to store validation-time mojo events
- Added captureValidationTimeProperties() to BuildCacheMojosExecutionStrategy
- Modified execute() to capture properties after findCachedBuild()
- Modified save() to store ONLY validation-time events (AssertionError if missing)
- Added comprehensive integration tests (Maven4Jpms, ExplicitModuleVersion, NonJpms, MultiModule)

Benefits:
- No configuration required (vs ignorePattern in PR #391)
- Fixes root cause (timing mismatch) not symptom (value mismatch)
- Works for ALL Maven 4 auto-injected properties
- Preserves execution-time logging for debugging

Alternative to PR #391

elharo reviewed (changes requested):
need to run mvn spotless:apply

cowwoc committed:
Apply code formatting and upgrade spotless tooling

- Upgrade spotless-maven-plugin from 2.44.5 to 3.0.0
- Upgrade palantir-java-format from 2.56.0 to 2.81.0 for Java 25 support
- Run mvn spotless:apply to format code

Addresses review feedback on PR #395.

cowwoc commented:
@elharo Done. Try now.

cowwoc committed:
Address PR #395 review feedback

- Split run-on sentence in AssertionError message
- Use HashMap import instead of fully qualified name
- Move Mojo variable declaration into try block (nested try-finally)
- Make getValidationTimeEvents() package-private

cowwoc commented on src/main/java/org/apache/maven/buildcache/BuildCacheMojosExecutionStrategy.java:
```diff
@@ -163,8 +176,16 @@ public void execute(
+                            || result.getValidationTimeEvents().isEmpty()) {
+                        throw new AssertionError(
+                                "Validation-time properties not captured for project " + projectName
+                                        + ". This is a bug - validation-time capture should always succeed when saving to cache.");
```
Hopefully I understood your comment correctly. See https://github.com/apache/maven-build-cache-extension/pull/395/commits/6cf3a361c988f9acdf97b291f93e8af9a6a0551f

cowwoc committed:
Fix test failures and improve validation-time capture

- Fix invalid XML in pom.xml comment (-- not allowed in XML comments)
- Fix cache config XML validation errors by simplifying config
- Change tests to use 'package' goal instead of 'compile' since cache
  extension stores artifacts, not intermediate output directories
- Fix AssertionError when running clean-only builds by gracefully
  skipping cache storage when no cacheable mojos executed

cowwoc committed:
Fix Spotless formatting violation

Put LOGGER.debug() call on single line to satisfy Spotless formatting rules.

erik-meuwese-topicus commented on src/test/projects/maven4-jpms-module/src/main/java/org/apache/maven/caching/test/maven4/HelloMaven4.java:
```diff
@@ -0,0 +1,7 @@
+package org.apache.maven.caching.test.maven4;
```
```suggestion
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.maven.caching.test.maven4;
```

The code convention requires the ASF license header
https://maven.apache.org/developers/conventions/code.html

erik-meuwese-topicus commented on src/test/projects/explicit-module-version/src/main/java/module-info.java:
```diff
@@ -0,0 +1,3 @@
+module org.apache.maven.caching.test.explicit {
```
```suggestion
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
module org.apache.maven.caching.test.explicit {
```

The code convention requires the ASF license header
https://maven.apache.org/developers/conventions/code.html

erik-meuwese-topicus commented on src/test/projects/maven4-jpms-module/.mvn/extensions.xml:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
```suggestion
<?xml version="1.0" encoding="UTF-8"?>
<!--
    Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to You under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

         http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.
-->
```
The code convention requires the ASF license header
https://maven.apache.org/developers/conventions/code.html

erik-meuwese-topicus commented on src/test/projects/non-jpms-project/.mvn/extensions.xml:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
```suggestion
<?xml version="1.0" encoding="UTF-8"?>
<!--
    Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to You under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

         http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.
-->
```
The code convention requires the ASF license header
https://maven.apache.org/developers/conventions/code.html

cowwoc committed:
Add ASF license headers to test project files

Add missing Apache Software Foundation license headers to all test project
files as required by Apache conventions:

- explicit-module-version: pom.xml, extensions.xml, module-info.java, ExplicitVersionModule.java
- maven4-jpms-module: pom.xml, extensions.xml, module-info.java, HelloMaven4.java
- multi-module-jpms: parent and module pom.xml files, extensions.xml, module-info.java, ModuleA.java, ModuleB.java
- non-jpms-project: pom.xml, extensions.xml, RegularJavaClass.java

maven-build-cache-config.xml files already had correct license headers.

cowwoc commented:
@elharo @olamy I've rebased the PR. Please check if the build passes now.

If it does, please consider merging since the PR is already marked as approved.

cowwoc commented:
@olamy please take another shot at this. It should be ready to merge.

cowwoc committed:
Apply Spotless formatting for cache storage log statement

cowwoc commented:
@olamy I fixed the Spotless check failure. Please trigger another build.

cowwoc commented:
@olamy @slachiewicz Can you please merge this PR?

cowwoc commented on src/main/java/org/apache/maven/buildcache/BuildCacheMojosExecutionStrategy.java:
```diff
@@ -183,9 +197,16 @@ public void execute(
                     } else {
-                        final Map<String, MojoExecutionEvent> executionEvents =
-                                mojoListener.getProjectExecutions(project);
-                        cacheController.save(result, mojoExecutions, executionEvents);
+                        // Only save cache if there are validation-time events to store
+                        // When running only clean phase, there are no cacheable mojos
+                        if (result.getValidationTimeEvents() == null
+                                || result.getValidationTimeEvents().isEmpty()) {
+                            LOGGER.debug("Skipping cache storage for {} - no cacheable mojos executed", projectName);
+                        } else {
+                            LOGGER.debug(
+                                    "Using validation-time properties for cache storage (consistent lifecycle point)");
+                            cacheController.save(result, mojoExecutions, result.getValidationTimeEvents());
+                        }
```
Addressed in f04d1cc. Validation-time capture raises LifecycleExecutionException for plugin-configuration failures, while the missing validation-event invariant raises AssertionError. The save path verifies that every cacheable mojo has a validation-time event before saving. Clean-only executions remain intentionally skipped. The PR description and code snippet match.

cowwoc replied:
Correction: the missing validation-event case is an internal invariant and now throws AssertionError; LifecycleExecutionException remains reserved for plugin-configuration failures. The PR description and code snippet have been updated accordingly.

cowwoc committed:
Fix validation-time property capture

cowwoc commented:
@cstamas Comment addressed. Please re-run the check.

cowwoc committed:
Address PR 395 review comments

cowwoc commented on src/main/java/org/apache/maven/buildcache/BuildCacheMojosExecutionStrategy.java:
```diff
@@ -444,6 +466,78 @@ boolean isParamsMatched(
+                    && lifecyclePhasesHelper.isLaterPhaseThanClean(mojoExecution.getLifecyclePhase())
+                    && (validationTimeEvents == null
+                            || !validationTimeEvents.containsKey(mojoExecutionKey(mojoExecution)))) {
+                throw new AssertionError("Validation-time properties not captured for project " + projectName);
```
Addressed in 8ba61a1: the assertion now includes the missing mojo’s full goal name and project name.

cowwoc commented on src/test/projects/explicit-module-version/pom.xml:
```diff
@@ -0,0 +1,57 @@
+        <version>3.13.0</version>
+        <configuration>
+          <!-- Explicit moduleVersion configuration - present at validation time -->
+          <compilerArgs>
+            <arg>--module-version</arg>
+            <arg>${project.version}</arg>
+            <arg>-parameters</arg>
+            <arg>-g</arg>
+          </compilerArgs>
```
Addressed in 8ba61a1: the test now configures <moduleVersion>${project.version}</moduleVersion> and leaves compilerArgs for the remaining compiler options.

cowwoc commented on src/main/java/org/apache/maven/buildcache/BuildCacheMojosExecutionStrategy.java:
```diff
@@ -134,6 +135,19 @@ public void execute(
+                                "Captured validation-time properties for {} mojos in project {}",
+                                validationTimeEvents.size(),
+                                projectName);
+                    }
```
Addressed in 8ba61a1: removed the redundant inner cacheState check so capture runs directly within the existing initialized block.

cowwoc commented on src/main/java/org/apache/maven/buildcache/BuildCacheMojosExecutionStrategy.java:
```diff
@@ -444,6 +466,78 @@ boolean isParamsMatched(
+            }
+
+            try {
+                mojoExecutionScope.enter();
```
Addressed in 8ba61a1: mojoExecutionScope.enter() now runs before the try/finally, so exit() is not attempted when enter() fails.

cowwoc commented:
@cstamas Round 2 complete :) Please re-check.

cowwoc committed:
Address PR 395 review feedback

cowwoc commented on src/main/java/org/apache/maven/buildcache/CacheResult.java:
```diff
@@ -111,4 +156,8 @@ public RestoreStatus getStatus() {
+    Map<String, MojoExecutionEvent> getValidationTimeEvents() {
+        return validationTimeEvents;
+    }
 }
```
Addressed in f4a1cf2: getValidationTimeEvents() is now public, matching the other CacheResult accessors and making the validation-time state usable by external callers.

cowwoc commented on src/main/java/org/apache/maven/buildcache/BuildCacheMojosExecutionStrategy.java:
```diff
@@ -444,6 +464,81 @@ boolean isParamsMatched(
+                            mojoExecution.getMojoDescriptor().getFullGoalName());
+                } finally {
+                    mavenPluginManager.releaseMojo(mojo, mojoExecution);
+                }
```
Addressed in f4a1cf2: validation-time Mojo instances are retained until cacheController.save() has reflected over them, then released in execute()'s finally block. Partial capture failures also release already-configured Mojos.

cowwoc commented:
@cstamas Please trigger another review.

cowwoc commented:
@cstamas The failure seems to be a transient network issue. Can you please trigger another build?

cstamas commented:
Please rebase against latest master.

cowwoc commented:
@cstamas Done. Try now.
